### ISSUE TITLE
Second generic should be added to HoCs to be able to specify return type of `props` callback.

### ISSUE THREAD

**Issue by RIP21** [dotansimha/graphql-code-generator#803]
Right now there is only one generic that gives the ability to specify `ownProps` but there is no optional second generic that will give the ability to specify `props` callback return type.

So the result will be like that:
```typescript
export function GetMessagesHOC<TProps = any, TChildProps = any>(
  operationOptions:
    | ReactApollo.OperationOption<
        TProps,
        GetMessagesQuery,
        GetMessagesVariables,
        Partial<ReactApollo.DataProps<GetMessagesQuery, GetMessagesVariables>> & TChildProps
      >
    | undefined,
) {
  return ReactApollo.graphql<
    TProps,
    GetMessagesQuery,
    GetMessagesVariables,
    Partial<ReactApollo.DataProps<GetMessagesQuery, GetMessagesVariables>> & TChildProps
  >(GetMessagesDocument, operationOptions)
}
```

So effectively `PropsCallback` optional generic added. And then `&` with `Partial` in both places.

**Comment by kamilkisiela** [comment]
Could you give an example how you might want to use it and where does it help you?

**Comment by RIP21** [comment]
It gives you confidence that you're return what you expect from prop
function, right now it's any type. Also it gives you ability to not mystype
keys etc.

czw., 8 lis 2018, 15:15: Kamil Kisiela <notifications@github.com>
napisał(a):

**Comment by RIP21** [comment]
So as an example. (I was using a phone)
```typescript
type PropsReturnType = {
  someValue: string
}

SomeHOC<{}, PropsReturnType>({
  props: () => ({
    somevalue: "Value", // This will error out since we did a typo, and should be someValue instead
    anotherValue: 25, // This will error out since we didn't provide such in PropsReturnType
    someValue: 30, // This will error out since someValue type is string.
  })
})
```

**Comment by dotansimha** [comment]
Fixed in 0.14.0 🎉
